src/Grpc.AspNetCore.Server/Internal/PipeExtensions.cs

```diff
@@ -317,6 +344,40 @@ private static bool TryReadMessage(ref ReadOnlySequence<byte> buffer, HttpContex
             var messageBuffer = buffer.Slice(HeaderSize, messageLength);
             message = messageBuffer.ToArray();
 
+            if (compressed)
+            {
+                string encoding = context.GetRequestGrpcEncoding();
+                if (encoding == null)
+                {
+                    throw new RpcException(NoMessageEncodingMessageStatus);
+                }
+                if (string.Equals(encoding, GrpcProtocolConstants.IdentityGrpcEncoding, StringComparison.Ordinal))
+                {
+                    throw new RpcException(IdentityMessageEncodingMessageStatus);
+                }
+
+                // Performance improvement would be to decompress without converting to an intermediary byte array
+                if (!GrpcProtocolHelpers.TryDecompressMessage(encoding, context.ServiceOptions.CompressionProviders, message, out var decompressedMessage))
+                {
+                    // https://github.com/grpc/grpc/blob/master/doc/compression.md#test-cases
+                    // A message compressed by a client in a way not supported by its server MUST fail with status UNIMPLEMENTED,
+                    // its associated description indicating the unsupported condition as well as the supported ones. The returned
+                    // grpc-accept-encoding header MUST NOT contain the compression method (encoding) used.
+                    var supportedEncodings = context.ServiceOptions.CompressionProviders.Select(p => p.EncodingName).ToList();
+
+                    if (!context.HttpContext.Response.HasStarted)
+                    {
+                        context.HttpContext.Response.Headers[GrpcProtocolConstants.MessageAcceptEncodingHeader] = string.Join(",", supportedEncodings);
+                    }
+
+                    throw new RpcException(CreateUnknownMessageEncodingMessageStatus(encoding, supportedEncodings));
+                }
+
+                context.ValidateAcceptEncodingContainsResponseEncoding();
```

Comment:
Curious why this check is needed?

Reply:
That's the second time you've asked that in this PR 😋